### ISSUE TITLE
[Request] BSNES enhancement features in SNES driver

### ISSUE THREAD

**Issue by Shideravan** [ares-emulator/ares#33]
Would be certainly nice to integrate all the specific SNES improvements which has in **bsnes**, such:

* HD Mode 7
* Personalize how many frames beyond in run-ahead mode
* Overclocking of each system/cartridge CPU
* Audio improvements
* Game enhancement
* Cheat support
* Widescreen support for modified games
* Compatibility options
* Game hotfixes

I'm thinking even in port myself some these things directly from **bsnes** code. There is any kind of restriction for this?

**Comment by Kawa-oneechan** [comment]
Why not just use bsnes if you want to use these features so much?

**Comment by Screwtapello** [comment]
bsnes is released under the GPLv3 licence; if those features were incorporated into ares, ares would inherit the GPLv3 licence instead of the ISC licence it uses now. The ares maintainers are pretty fond of the ISC licence, so I don't see this happening any time soon.

That said, the ares SNES core is probably *structurally* similar to the bsnes core (in that they both have scanline renderers, they both probably collect the results of each scanline before rendering anything, etc.) and the techniques bsnes uses are well known and understood, so it might not be too difficult to re-implement them in ares. The difficult thing would probably be testing - bsnes has had years of testing on all kinds of games to shake the bugs out; it will be difficult to get that amount of testing in ares unless it can become as popular as bsnes is/was.

**Comment by LukeUsher** [comment]
As screwtape specified, there are licensing concerns with backporting code from bsnes, ares is ISC licensed and bsnes is GPLv3 meaning that it is impossible to legally add code from bsnes to the ares project.

additionally, the following features are out-of-scope for ares:
 * Overclocking - ares (mostly) follows the accuracy goal of higan, with the caveat that accuracy is relaxed for newer systems that are impossible to emulate with the same approach. Overclocking contradicts this goal.
 * Compatibility options - these should not be required, as ares does not take the same shortcuts as bsnes does.

Can you elaborate on audio improvements and game hotfixes? ares already has support for MSU-1.

Widescreen support is partially implemented for SNES games, but currently, there is no way to enable it: this comes down to this feature being unfinished when Near made their final release.

**Comment by Screwtapello** [comment]
> Compatibility options - these should not be required, as ares does not take the same shortcuts as bsnes does

A minor correction: bsnes' "compatibility" options are for compatibility with romhacks designed for less-accurate emulators, not for covering up inaccurate emulation. Things like "allow games to write to VRAM during active display", that kind of thing.

> audio improvements

The original SNES used Gaussian interpolation when pitch-shifting audio samples (I think?) which was cheap to implement but muddies the audio quality somewhat. bsnes allows Cubic interpolation to be used instead, which makes the audio much clearer, but is technically inaccurate.

> hotfixes

bsnes' "hotfixes" are basically "don't randomise the initial contents of RAM for games that can't handle it" (and would therefore likely be buggy on real consoles too):

https://github.com/bsnes-emu/bsnes/blob/063e645ff27faba651c80b9f4648b91d2b85a79c/bsnes/target-bsnes/program/hacks.cpp#L49-L57

**Comment by LukeUsher** [comment]
> > Compatibility options - these should not be required, as ares does not take the same shortcuts as bsnes does
> 
> A minor correction: bsnes' "compatibility" options are for compatibility with romhacks designed for less-accurate emulators, not for covering up inaccurate emulation. Things like "allow games to write to VRAM during active display", that kind of thing.
> 
> > audio improvements
> 
> The original SNES used Gaussian interpolation when pitch-shifting audio samples (I think?) which was cheap to implement but muddies the audio quality somewhat. bsnes allows Cubic interpolation to be used instead, which makes the audio much clearer, but is technically inaccurate.
> 
> > hotfixes
> 
> bsnes' "hotfixes" are basically "don't randomise the initial contents of RAM for games that can't handle it" (and would therefore likely be buggy on real consoles too):
> 
> https://github.com/bsnes-emu/bsnes/blob/063e645ff27faba651c80b9f4648b91d2b85a79c/bsnes/target-bsnes/program/hacks.cpp#L49-L57

Thanks for the corrections, but I will still maintain the view that these features are better suited for bsnes, the compatibility options especially sound like they're out of scope for this project.

**Comment by LukeUsher** [comment]
Marked as wontfix for now, reasoning above;

**Comment by Shideravan** [comment]
Thanks for all the clarification @LukeUsher and @Screwtapello. These are really strong arguments to not just "port" these features directly from BSNES. I'm happy to have opened this issue before starting working to port those features to ares.
At the same time, I think that there's no restriction about create a fork from ares to merge these feature and distribute them as GPLv3, am I right?
That said, it doesn't mean that I do this. At least not in the short term, as I would be more interested in using my free time on things that are really useful in the main fork of ares (or in other emulators projects in general.
This issue can be safely closed.

**Comment by qwertymodo** [comment]
Another potential option for porting features would be to approach the authors of those specific features (since a lot of those things were created individually by a single author each) about dual/re-licensing the feature as ISC for ares. The more core integrated bits of the application are more difficult in that regard, but standalone features might be possible.
